torchaudio/datasets/speechcommands.py

```diff
@@ -79,9 +92,22 @@ def __init__(self,
                     download_url(url, root, hash_value=checksum, hash_type="md5")
                 extract_archive(archive, self._path)
 
-        walker = walk_files(self._path, suffix=".wav", prefix=True)
-        walker = filter(lambda w: HASH_DIVIDER in w and EXCEPT_FOLDER not in w, walker)
-        self._walker = list(walker)
+        def load_list(filename):
+            filepath = os.path.join(self._path, filename)
+            with open(filepath) as f:
+                return [os.path.join(self._path, l.strip()) for l in f]
+
+        if subset == "validation":
+            self._walker = load_list(VALIDATION_LIST)
+        elif subset == "testing":
+            self._walker = load_list(TESTING_LIST)
+        elif subset == "training":
+            excludes = load_list(VALIDATION_LIST) + load_list(TESTING_LIST)
+            walker = walk_files(self._path, suffix=".wav", prefix=True)
+            self._walker = [w for w in walker if HASH_DIVIDER in w and EXCEPT_FOLDER not in w and w not in excludes]
```

Comment:
Looking way better than the previous. One more push.

Rather than parsing the all the files present in the file system first then filter only the interested, it's better to only parse the needed one.

Also, repeating the combination of `walk_files` and filtering `HASH_DIVIDER` in two clauses looks overwhelming. You can combine them into a much shorter code like the bellow.

```python
elif subset == "testing":
    ...
else:
    paths = [str(p) for p in Path(self._path).glob('*/*_no_hash_*.wav')]
        # no need to check EXCEPT_FOLDER because
        # none of the wav file in the directory match with `_no_hash_`
    if subset == "training":
        excludes = set(_load_lists("validation_list.txt", "testing_list.txt"))
        paths = [p for p in paths if p not in excludes]
    self._walker = paths

Reply:
I agree that parsing and then filtering is not optimal. Changing `os_walk` is out of scope for this PR though, and could change the order of the files from #814. I'm ok with replacing `os_walk` with `glob` in a next PR due to #950.

Reply:
I encourage you to take this opportunity ro get rid of `walk_files` because, if you try to stick with `walk_files`, I do not see any pro at a glance. You will have to more works to settle with sub-optimal solution, which is not as compact yet inefficient.

Also #950 is about the support of `Path` object, so I do not see the logical connection between it and `walk_files` (which returns list of string).

Regarding, the order it returns, we have a unit test which can verify the proper ordering of files returned, so you do not need to worry about aligning the order with `walk_files` function. Instead, you can just focus on writing the correct code. However, this is a valid point and probably you will need to insert `sort` function in the snippet above.

Reply:
@vincentqb Now that #950 is closed, can you followup replacing `walk_files` with `glob` with something like the following?

> I agree that parsing and then filtering is not optimal. Changing `os_walk` is out of scope for this PR though, and could change the order of the files from #814. I'm ok with replacing `os_walk` with `glob` in a next PR due to #950.

> ```python
> elif subset == "testing":
>     ...
> else:
>     paths = [str(p) for p in Path(self._path).glob('*/*_no_hash_*.wav')]
>         # no need to check EXCEPT_FOLDER because
>         # none of the wav file in the directory match with `_no_hash_`
>     if subset == "training":
>         excludes = set(_load_lists("validation_list.txt", "testing_list.txt"))
>         paths = [p for p in paths if p not in excludes]
>     self._walker = paths
> ```

Reply:
(about #940: #950 discussed using the Path library which arrived in 3.4. now that we use it, i'm ok with using other functions from it.)

Reply:
good idea: i've opened an issue to request help from the community to remove `walk_files`: #1051.